# Immunotherapy Microbiome Analysis (PRJEB61942) 

This repository contains all scripts and ressources related to the analysis of the PRJEB61942 dataset, which investigates the link between the microbiome and response to immunotherapy in melanoma patients.


## Workflow: 

### 1. Data selection
From the ENA metadata, I selected only the **16S rRNA** samples based on the "tax_id=9606" and confirmed the sequencing method 

### 2. Downloading the FASTQ files 
Using a script ("downlaod_fastq.sh"), I downloaded all paired-end FASTQ files related to 16S sequencing from the ENA FTP links

### 3. Files integrity check 
To make sure files were correctly downloaded: 
- I used the **md5 checksums** provided in the metadata
- I wrote a bash script ("verify_md5.sh") to compare them with local files
- A log of passed and failed checks was created
  
### 4. Quality control with fastp 
I ran fastp on all paired-end FASTQ files using default settings 
It cleaned low-quality reads and generated JSON and HTML reports for each sample. 

### 5. Quality summary and filtering decision
An R script ("scripts/fastp_summary.r") was created to parse the JSON reports generated by fastp.
It extracts important metrics (number of reads before and after filtering, Q20 and Q30 rates), summarizes them into a tale and produces multiple visualizations. 
**Key observations:**
* Q20 and Q30 scores remain high for all samples
* The main filtering criterion was he number of reads after filtering.

So finally, only samples with less than 30,000 reads after filtering were excluded from downstream analysis.
A detailed visual report explaining this is available: "metadata/fastp_summary_report.pdf" 

### 6. DADA2
The DADA2 algorithm was used to process the 16S rRNA sequencing data. It allows high resolution inference of Amplicon Sequence Variants (ASVs) by correcting sequencing errors without clustering into OTUs.
We followed the official [DADA2 tutorial](https://benjjneb.github.io/dada2/tutorial.html) as a guideline, adapting it to our dataset and processing contraints. 

#### 6.1 DADA2 analysis - Manual version: 
As a first step, DADA2 was run manually on a subset of filtered samples to explore and validate each part of the workflow. This was done using a single R script: 
'scripts/dada2.R'

The main steps performed were: 
- **filtering and trimming** with 'filterAndTrim()'
- **learning error rates** using 'learnErrors()'
- **dereplication** with 'derepFastq()'
- **ASV inference** using 'dada()'
- **Merging paired-end reads** via 'mergePairs()
- **sequence table construction** with 'makeSequenceTable()'
- **Chimera removal** using 'removeBimeraDenovo()'
- **Taxonomic assignement** using SILVA reference databases with 'assignTaxonomy()' and 'addSpecies()'

This setp-by-step approach allowed to verify that each DADA2 step ran correctly before developing a full automated pipeline.

#### 6.1 DADA2 analysis - Automated pipeline: 

To streamline the process and make it reusable on other datasets, the DADA2 analysis was restructured into a **modular and automated pipeline** using multiple R scripts, coordinated through a bash script: 
'scripts/dada2_automated/dada2_auto_pipeline.sh' 

## File structure: 

### scripts:
* download_fastq.sh    #download FATSQ files from ENA
* verify_md5.sh        #check the md5 integrity of downloaded files
* run_fastp.sh         #quality filtering with fastp
* fastp_summary.r      #generate plots and summary from fastp JSON reports 

### metadata: 
* filereport_16S.txt   #cleaned metadata with 16S samples
* fastq_links_16S.txt  #list of FASTQ URLs to download
* filereport_16S_filtered.tsv  #metadata filtered to keep only good quality samples
* fastp_summary_table.csv   #CSV table with reads and quality scores

### logs: 
* md5_check_ok.txt     #files that passed the md5 check
* md5_check_fail.txt   #files that failed or were missing
* log_download.txt     #output log of the download script

### others
* fastp_summary_report.pdf #visual explaination of filtering decisions 


